core/network/fabric-interop-cc/contracts/interop/write_external_state.go

```diff
@@ -52,40 +64,61 @@ func ExtractAndValidateDataFromView(view *common.View, b64ViewContent string) ([
 		if err != nil {
 			return nil, fmt.Errorf("Unable to Unmarshal interopPayload: %s", err.Error())
 		}
+		interopPayloadList[0] = &interopPayload
 	} else {
 		return nil, fmt.Errorf("Cannot extract data from view; unsupported DLT type: %+v", view.Meta.Protocol)
 	}
 
-	// If view data is encrypted, match it to supplied decrypted data using the hash in the view payload
-	if interopPayload.Confidential {
-		var confidentialPayload common.ConfidentialPayload
-		err := protoV2.Unmarshal(interopPayload.Payload, &confidentialPayload)
-		if err != nil {
-			return nil, fmt.Errorf("ConfidentialPayload Unmarshal error: %s", err)
-		}
-		viewB64ContentBytes, err := base64.StdEncoding.DecodeString(b64ViewContent)
-		if err != nil {
-			return nil, fmt.Errorf("Unable to base64 decode view content: %s", err.Error())
-		}
-		var confidentialPayloadContents common.ConfidentialPayloadContents
-		err = protoV2.Unmarshal(viewB64ContentBytes, &confidentialPayloadContents)
-		if err != nil {
-			return nil, fmt.Errorf("ConfidentialPayloadContents Unmarshal error: %s", err)
-		}
-		if confidentialPayload.HashType == common.ConfidentialPayload_HMAC {
-			payloadHMAC := hmac.New(sha256.New, confidentialPayloadContents.Random)
-			payloadHMAC.Write(confidentialPayloadContents.Payload)
-			payloadHMACBytes := payloadHMAC.Sum(nil)
-			if !bytes.Equal(confidentialPayload.Hash, payloadHMACBytes) {
-				return nil, fmt.Errorf("View payload hash does not match hash of data submitted by client")
+	var payloadConfidential bool
+	var viewPayload []byte
+	var confidentialPayloadContents common.ConfidentialPayloadContents
+	for i, interopPayload := range interopPayloadList {
+		// If view data is encrypted, match it to supplied decrypted data using the hash in the view payload
+		if interopPayload.Confidential {
+			if i == 0 {
+				payloadConfidential = true
+				// Unmarshal the (decrypted) confidential payload contents supplied by the caller; needs to be done only once
+				viewB64ContentBytes, err := base64.StdEncoding.DecodeString(b64ViewContent)
+				if err != nil {
+					return nil, fmt.Errorf("Unable to base64 decode view content: %s", err.Error())
+				}
+				err = protoV2.Unmarshal(viewB64ContentBytes, &confidentialPayloadContents)
+				if err != nil {
+					return nil, fmt.Errorf("ConfidentialPayloadContents Unmarshal error: %s", err)
+				}
+				viewPayload = confidentialPayloadContents.Payload
+            } else if !payloadConfidential {
+				return nil, fmt.Errorf("Mismatching confidentiality flags among interop payloads")
+			}
+			var confidentialPayload common.ConfidentialPayload
+			err := protoV2.Unmarshal(interopPayload.Payload, &confidentialPayload)
+			if err != nil {
+				return nil, fmt.Errorf("ConfidentialPayload Unmarshal error: %s", err)
+			}
+			if confidentialPayload.HashType == common.ConfidentialPayload_HMAC {
+				payloadHMAC := hmac.New(sha256.New, confidentialPayloadContents.Random)
```

Comment:
Assuming every peer/node generates different random `r`, how can one decrypted payload's random value `r` verify other peer/node's HMAC hash which may have used different `r`?

Reply:
Don't we need all peers' decrypted payloads? assuming each one will have different `r`.

Reply:
Excellent point!

Reply:
Fixed logic in both chaincode and SDK to use and validate multiple decrypted payloads with different randomness values.